File: services/php-laravel/app/Http/Controllers/InteractionController.php
<?php

namespace App\Http\Controllers;

use App\Services\PubSubPublisher;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;

class InteractionController
{
    // Interaction types
    private const INTERACTION_TYPE_PING = 1;

    private const INTERACTION_TYPE_APPLICATION_COMMAND = 2;

    // Response types
    private const RESPONSE_TYPE_PONG = 1;

    private const RESPONSE_TYPE_DEFERRED_CHANNEL_MESSAGE = 5;

    private PubSubPublisher $pubsub;

    public function __construct(PubSubPublisher $pubsub)
    {
        $this->pubsub = $pubsub;
    }

    public function handle(Request $request): JsonResponse
    {
        $interaction = $request->json()->all();

        // Ensure interaction is valid
        if (empty($interaction) || ! is_array($interaction)) {
            return response()->json(['error' => 'invalid JSON'], 400);
        }

        $type = $interaction['type'] ?? null;

        // Check if type exists
        if ($type === null) {
            return response()->json(['error' => 'unsupported interaction type'], 400);
        }

        // Check if type is an integer
        if (! is_int($type)) {
            return response()->json(['error' => 'unsupported interaction type'], 400);
        }

        return match ($type) {
            self::INTERACTION_TYPE_PING => $this->handlePing(),
            self::INTERACTION_TYPE_APPLICATION_COMMAND => $this->handleApplicationCommand($interaction),
            default => response()->json(['error' => 'unsupported interaction type'], 400),
        };
    }

    private function handlePing(): JsonResponse
    {
        // Respond with Pong - do NOT publish to Pub/Sub
        return response()->json(['type' => self::RESPONSE_TYPE_PONG]);
    }

    private function handleApplicationCommand(array $interaction): JsonResponse
    {
        // Publish to Pub/Sub asynchronously (in PHP we do it inline but quickly)
        $this->pubsub->publish($interaction);

        // Respond with deferred response (non-ephemeral)
        return response()->json(['type' => self::RESPONSE_TYPE_DEFERRED_CHANNEL_MESSAGE]);
    }
}

File: services/scala-play/app/controllers/InteractionController.scala
package controllers

import org.apache.pekko.util.ByteString
import play.api.libs.json.*
import play.api.mvc.*
import services.{PubSubService, SignatureService}

import javax.inject.*
import scala.concurrent.{ExecutionContext, Future}
import scala.util.{Failure, Success, Try}

/** Discord interaction controller.
  *
  * Handles Discord webhook interactions:
  *   - Validates Ed25519 signatures
  *   - Responds to Ping (type=1) with Pong (type=1)
  *   - Responds to Slash commands (type=2) with Deferred (type=5)
  *   - Publishes sanitized slash command payloads to Pub/Sub
  */
@Singleton
class InteractionController @Inject() (
  val controllerComponents: ControllerComponents,
  signatureService: SignatureService,
  pubSubService: PubSubService
)(implicit ec: ExecutionContext)
    extends BaseController:

  // Interaction types
  private val InteractionTypePing = 1
  private val InteractionTypeApplicationCommand = 2

  // Response types
  private val ResponseTypePong = 1
  private val ResponseTypeDeferredChannelMessage = 5

  /** Health check endpoint. */
  def health: Action[AnyContent] = Action {
    Ok(Json.obj("status" -> "ok"))
  }

  /** Handle Discord interaction webhook. */
  def handleInteraction: Action[RawBuffer] = Action(parse.raw) { request =>
    // Get raw body for signature verification
    val bodyBytes = request.body.asBytes().getOrElse(ByteString.empty)
    val bodyString = bodyBytes.utf8String

    // Get signature headers
    val signature = request.headers.get("X-Signature-Ed25519").getOrElse("")
    val timestamp = request.headers.get("X-Signature-Timestamp").getOrElse("")

    // Validate signature
    if !signatureService.validateSignature(signature, timestamp, bodyString) then
      Unauthorized(Json.obj("error" -> "invalid signature"))
    else
      // Parse interaction - handle malformed JSON gracefully
      Try(Json.parse(bodyString)) match
        case Success(json) =>
          json.validate[JsObject] match
            case JsSuccess(interaction, _) =>
              handleInteractionByType(interaction)
            case JsError(_) =>
              BadRequest(Json.obj("error" -> "invalid JSON"))
        case Failure(_) =>
          BadRequest(Json.obj("error" -> "invalid JSON"))
  }

  private def handleInteractionByType(interaction: JsObject): Result =
    (interaction \ "type").asOpt[Int] match
      case Some(InteractionTypePing) =>
        handlePing()
      case Some(InteractionTypeApplicationCommand) =>
        handleApplicationCommand(interaction)
      case _ =>
        BadRequest(Json.obj("error" -> "unsupported interaction type"))

  /** Handle Ping interaction - respond with Pong. Do NOT publish to Pub/Sub. */
  private def handlePing(): Result =
    Ok(Json.obj("type" -> ResponseTypePong))

  /** Handle Application Command (slash command) interaction. */
  private def handleApplicationCommand(interaction: JsObject): Result =
    // Publish to Pub/Sub in background (if configured)
    Future {
      pubSubService.publish(interaction)
    }

    // Respond with deferred response (non-ephemeral)
    Ok(Json.obj("type" -> ResponseTypeDeferredChannelMessage))
